app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt

```diff
@@ -0,0 +1,280 @@
+        }
+
+        responseObserver.onNext(createPrivateKey())
+        responseObserver.onCompleted()
+    }
+
+    fun saveBlock(request: ChainService.PeerTrustChainBlock): ValidationResult? = saveBlock(request.peer, request.block)
+
+    fun saveBlock(peer: ChainService.Peer, block: MessageProto.TrustChainBlock): ValidationResult? {
+        Log.i(TAG, "Received half block from peer with IP: " + peer.hostname + ":" + peer.port +
+                " and public key: " + Peer.bytesToHex(peer.publicKey.toByteArray()))
+
+        val validation: ValidationResult
+        try {
+            validation = TrustChainBlock.validate(block, dbHelper)
+        } catch (e: Exception) {
+            e.printStackTrace()
+            return null
+        }
+
+        Log.i(TAG, "Received block validation result " + validation.toString() + "("
+                + TrustChainBlock.toString(block) + ")")
+
+        if (validation.getStatus() === ValidationResult.ValidationStatus.INVALID) {
+            for (error in validation.getErrors()) {
+                Log.e(TAG, "Validation error: " + error)
+            }
+            return null
+        } else {
+            dbHelper.insertInDB(block)
+        }
+
+        val pk = keyPair.public.encoded
+        // check if addressed to me and if we did not sign it already, if so: do nothing.
+        if (block.linkSequenceNumber != TrustChainBlock.UNKNOWN_SEQ ||
+                !Arrays.equals(block.linkPublicKey.toByteArray(), pk) ||
+                null != dbHelper.getBlock(block.linkPublicKey.toByteArray(),
+                        block.linkSequenceNumber)) {
+            Log.e(TAG, "Received block not addressed to me or already signed by me.")
+            return null
+        }
+
+        // determine if we should sign the block, if not: do nothing
+        if (!shouldSign(block)) {
+            Log.e(TAG, "Will not sign received block.")
+            return null
+        }
+
+        return validation
+    }
+
+    fun connectToPeer(peer: PeerItem): List<ChainService.PeerTrustChainBlock> {
+        val connectablePeer = peer.asPeerMessage()
+        val crawledBlocks = sendCrawlRequest(connectablePeer, myPublicKey, -5)
+
+        val bufferedCrawledBlocks = crawledBlocks.asSequence().toList()
+
+        for (crawledBlock in bufferedCrawledBlocks) {
+            saveBlock(crawledBlock)
+        }
+
+        return bufferedCrawledBlocks
+    }
+
+    fun sendBlockToKnownPeer(peer: PeerItem, payload: String): ChainService.PeerTrustChainBlock? {
+        val sequenceNumberForCrawl = sequenceNumberForCrawl(-5)
+        val crawledBlocks = dbHelper.crawl(myPublicKey, sequenceNumberForCrawl)
+
+        val crawlResponse = ChainService.CrawlResponse.newBuilder().setPeer(me).addAllBlock(crawledBlocks).build()
+        val peerMessage = peer.asPeerMessage()
+        val peerChannel = registry.channelForPeer(peerMessage)
+        val theirPublicKey: ChainService.Key = peerChannel.sendLatestBlocks(crawlResponse)
```

Comment:
sendLatestBlocks returns the other party's public key?

Reply:
Yup, a bit weird, this was before we had the getPublicKey message, we could remove this but it saves a message